librad/src/git/replication.rs

```diff
@@ -102,198 +136,623 @@ pub fn replicate<Addrs>(
+            })
+            .map_err(|e| Error::Fetch(e.into()))
+            .and_then(project::fetched_peers)?;
+
+        let remote_ident =
+            unsafe_into_urn(Reference::rad_id(Namespace::from(&urn)).with_remote(remote_peer));
+        Ok(Replication::Clone {
+            urn,
+            fetched_peers,
+            identity: identities::any::get(storage, &remote_ident)?
+                .ok_or(Error::MissingIdentity)?,
+        })
+    } else {
+        let identity = identities::any::get(storage, &urn)?.ok_or(Error::MissingIdentity)?;
+        let existing = match identity {
+            SomeIdentity::Project(ref proj) => {
+                let mut remotes = project::all_delegates(&proj);
+                let mut tracked = tracking::tracked(storage, &urn)?.collect::<BTreeSet<_>>();
+                remotes.append(&mut tracked);
+
+                remotes
+            },
+            SomeIdentity::Person(_) => tracking::tracked(storage, &urn)?.collect::<BTreeSet<_>>(),
+        };
+
+        let _ = fetcher
+            .fetch(fetch::Fetchspecs::Peek {
+                remotes: existing.clone(),
+                max_fetch: fetch::ONE_MB * existing.len(),
+            })
+            .map_err(|e| Error::Fetch(e.into()))?;
+
+        Ok(Replication::Fetch {
+            urn,
+            identity,
+            existing,
+        })
     }
 }
 
+fn unsafe_into_urn(reference: Reference<git_ext::RefLike>) -> Urn {
+    reference.try_into().expect("namespace is set")
+}
+
+#[allow(clippy::unit_arg)]
 #[tracing::instrument(level = "trace", skip(storage), err)]
-fn ensure_setup_as_project(
+fn ensure_rad_id(storage: &Storage, urn: &Urn, tip: ext::Oid) -> Result<(), Error> {
+    identities::common::IdRef::from(urn)
+        .create(storage, tip)
+        .map_err(|e| Error::Store(e.into()))
+}
+
+/// Untrack the list of `PeerId`s, which also has the side-effect of removing
+/// that peer's remote references in the storage.
+///
+/// In the case of an error occurring during the untrack call we emit a warning
+/// and append it to a `Vec` of unpruned peers. The unpruned peers are returned.
+#[allow(clippy::unit_arg)]
+#[tracing::instrument(level = "trace", skip(storage, prune_list))]
+fn prune<'a>(
     storage: &Storage,
-    proj: Project,
-    remote_peer: PeerId,
-) -> Result<impl Iterator<Item = Urn>, Error> {
-    let local_peer_id = storage.peer_id();
+    urn: &Urn,
+    prune_list: impl Iterator<Item = &'a PeerId>,
+) -> Vec<PeerId> {
+    let mut unpruned = vec![];
+
+    for peer in prune_list {
+        match tracking::untrack(storage, urn, *peer) {
+            Ok(removed) => {
+                if removed {
+                    tracing::info!("pruned `{}`", peer);
+                } else {
+                    tracing::trace!("attempted to prune `{}` but it did not exist", peer);
+                }
+            },
+            Err(err) => {
+                tracing::warn!("failed to prune `{}`\nreason: {}", peer, err);
```

Comment:
I think the loop should be aborted on the first error, as it's somewhat unlikely that the next iteration will succeed.

Also nit: newlines in log messages are probably not improving readability. I've settled on `xyz!(err = ?e, "somethingsomething error")` elsewhere.